ibverbs/src/lib.rs

```diff
@@ -368,33 +372,53 @@ impl Context {
     /// A protection domain is a means of protection, and helps you create a group of object that
     /// can work together. If several objects were created using PD1, and others were created using
     /// PD2, working with objects from group1 together with objects from group2 will not work.
-    pub fn alloc_pd(&self) -> Result<ProtectionDomain, ()> {
+    pub fn alloc_pd(self: Arc<Self>) -> Result<ProtectionDomain, ()> {
         let pd = unsafe { ffi::ibv_alloc_pd(self.ctx) };
         if pd.is_null() {
             Err(())
         } else {
-            Ok(ProtectionDomain { ctx: self, pd })
+            Ok(ProtectionDomain {
+                ctx: self.clone(),
+                pd,
+            })
+        }
+    }
+
+    /// Query corresponding device
+    pub fn query_device(&self) -> io::Result<DeviceAttr> {
+        let mut dev_attrs = Default::default();
+
+        let ret = unsafe { ffi::ibv_query_device(self.ctx, &mut dev_attrs) };
+
+        if ret != 0 {
+            return Err(io::Error::last_os_error());
         }
+
+        Ok(dev_attrs)
     }
 }
 
 impl Drop for Context {
     fn drop(&mut self) {
-        let ok = unsafe { ffi::ibv_close_device(self.ctx) };
+        let ok = if !self.ctx.is_null() {
+            unsafe { ffi::ibv_close_device(self.ctx) }
+        } else {
+            0
+        };
         assert_eq!(ok, 0);
     }
 }
 
 /// A completion queue that allows subscribing to the completion of queued sends and receives.
-pub struct CompletionQueue<'ctx> {
-    _phantom: PhantomData<&'ctx ()>,
+pub struct CompletionQueue {
+    // _phantom: PhantomData<&'ctx ()>,
```

Comment:
I was under the impression that a CQ _must_ outlive the context — is that not right?

Reply:
Let me double check this one too. The changes on this file are mostly from nevenmoo, so this might just have been them getting around lifetime requirements for ease of use.

Reply:
Should we tag them in to this PR? They may be able to provide some additional input.

Reply:
Oh! I did, it's @nevenoomo. With another tag they might see it :P